src/celeritas/global/ActionLauncher.device.hh

```diff
@@ -32,24 +32,68 @@ namespace
 {
 //---------------------------------------------------------------------------//
 /*!
- * Launch the given executor using thread ids in the thread_range
+ * Celeritas executor kernel implementation.
  */
+
 template<class F>
-__global__ void
-launch_action_impl(Range<ThreadId> const thread_range, F execute_thread)
+__device__ CELER_FORCEINLINE void
+launch_kernel_impl(Range<ThreadId> const thread_range, F& execute_thread)
 {
     auto tid = celeritas::KernelParamCalculator::thread_id();
     if (!(tid < thread_range.size()))
         return;
     execute_thread(*(thread_range.cbegin() + tid.get()));
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Launch the given executor using thread ids in the thread_range
+ */
+
+template<class F>
+__global__ void
+launch_action_impl(Range<ThreadId> const thread_range, F execute_thread)
+{
+    launch_kernel_impl(thread_range, execute_thread);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Launch the given executor using thread ids in the thread_range with
+ \c __launch_bounds__
+ */
+
+#if CELERITAS_USE_CUDA
+template<class F, int T, int B = 1, int B_FINAL = B>
+#elif CELERITAS_USE_HIP
+// see
+// https://rocm.docs.amd.com/projects/HIP/en/latest/reference/kernel_language.html#porting-from-cuda-launch-bounds
+template<class F, int T, int B = 1, int B_FINAL = (B * T) / 32>
```

Comment:
I agree with the HIP documentation that specifying the second parameter as "warps per execution unit" is more sensible than "minimum number of threads". That value more directly corresponds to maximum occupancy. Am I crazy? If not then let's change the second parameter to be HIP-native and convert to "min threads per SM" for CUDA.

Reply:
Yeah I can see the HIP version being more sensible though I'd assume more people are familiar with the CUDA semantics. I'll change that to use the HIP version.